### ISSUE TITLE
The set() statement in gp.cxOnePoint cause reproducibility issue with a fixed random seed

### ISSUE THREAD

**Issue by weixuanfu2016** [DEAP/deap#190]
Currently when using DEAP with given random seed, not always the same results are obtained. After digging into the codes, I and @rhiever found that it caused by `gp.cxOnePoint`

Below is the codes (based on spambase example) for reproducing the issue.

```
from deap import gp
from deap import creator
from deap import base
import random
import operator
import itertools
from collections import defaultdict

for i in range(10):
    random.seed(42)

    class bar:
        pass

    class foo:
        pass

    # defined a new primitive set for strongly typed GP
    pset = gp.PrimitiveSetTyped("MAIN", itertools.repeat(bar, 57), foo, "IN")

    # boolean operators
    pset.addPrimitive(operator.and_, [foo, foo], foo)
    pset.addPrimitive(operator.or_, [foo, foo], foo)
    pset.addPrimitive(operator.not_, [foo], foo)

    # floating point operators
    # Define a protected division function
    def protectedDiv(left, right):
        try: return left / right
        except ZeroDivisionError: return 1

    pset.addPrimitive(operator.add, [bar,bar], bar)
    pset.addPrimitive(operator.sub, [bar,bar], bar)
    pset.addPrimitive(operator.mul, [bar,bar], bar)
    pset.addPrimitive(protectedDiv, [bar,bar], bar)

    # logic operators
    # Define a new if-then-else function
    def if_then_else(input, output1, output2):
        if input: return output1
        else: return output2

    pset.addPrimitive(operator.lt, [bar, bar], foo)
    pset.addPrimitive(operator.eq, [bar, bar], foo)
    pset.addPrimitive(if_then_else, [foo, bar, bar], bar)

    # terminals
    pset.addTerminal(False, foo)
    pset.addTerminal(True, foo)

    creator.create("FitnessMax", base.Fitness, weights=(1.0,))
    creator.create("Individual", gp.PrimitiveTree, fitness=creator.FitnessMax)

    # just copy the code from gp.cxOnePoint
    # with printing out the order of list(common_types)
    def cxOnePoint(ind1, ind2):
        """Randomly select in each individual and exchange each subtree with the
        point as root between each individual.
        :param ind1: First tree participating in the crossover.
        :param ind2: Second tree participating in the crossover.
        :returns: A tuple of two trees.
        """
        # Define the name of type for any types.
        __type__ = object

        if len(ind1) < 2 or len(ind2) < 2:
            # No crossover on single node tree
            return ind1, ind2

        # List all available primitive types in each individual
        types1 = defaultdict(list)
        types2 = defaultdict(list)
        if ind1.root.ret == __type__:
            # Not STGP optimization
            types1[__type__] = range(1, len(ind1))
            types2[__type__] = range(1, len(ind2))
            common_types = [__type__]
        else:
            for idx, node in enumerate(ind1[1:], 1):
                types1[node.ret].append(idx)
            for idx, node in enumerate(ind2[1:], 1):
                types2[node.ret].append(idx)

            common_types = set(types1.keys()).intersection(set(types2.keys()))
            ##### print out
            print(list(common_types))


        if len(list(common_types)) > 0:
            type_ = random.choice(list(common_types))

            index1 = random.choice(types1[type_])
            index2 = random.choice(types2[type_])

            slice1 = ind1.searchSubtree(index1)
            slice2 = ind2.searchSubtree(index2)
            ind1[slice1], ind2[slice2] = ind2[slice2], ind1[slice1]

        return ind1, ind2

    ind1 = creator.Individual.from_string('lt(sub(IN14, IN48), if_then_else(False, IN21, IN38))', pset)
    ind2 = creator.Individual.from_string('eq(if_then_else(False, IN44, IN30), sub(IN39, IN45))', pset)
    new_ind1, new_ind2 = cxOnePoint(ind1, ind2)
    print(str(new_ind1), str(new_ind2))

```

A possible fix is just to change `common_types = set(types1.keys()).intersection(set(types2.keys()))` to `common_types = [x for x in list(types1.keys()) if x in list(types2.keys())]` or `common_types = [x for x in types1 if x in types2]`

**Comment by weixuanfu2016** [comment]
The issue is related to the unordered data structure `set()`. More issue can be found in this [link](http://stackoverflow.com/questions/2594920/behavior-of-object-in-set-operations)

**Comment by rhiever** [comment]
To clarify: The issue is that when using custom classes as types in DEAP, the `cxOnePoint` function does not produce the same results when the same random seed is set before using it. This issue arises because of the following line in `cxOnePoint`:

```Python
common_types = set(types1.keys()).intersection(set(types2.keys()))
```
because the `set` will order the class types differently based on their memory address, and the classes are assigned different addresses every time they are defined.

Our recommendation (and how we fixed it in [TPOT](https://github.com/rhiever/tpot)) is to not use set operations and use the code that @weixuanfu2016 posted above:

```Python
common_types = [x for x in types1 if x in types2]
```

**Comment by weixuanfu2016** [comment]
Additionally, [the same line](https://github.com/DEAP/deap/blob/master/deap/gp.py#L712) can be found in `gp.cxOnePointLeafBiased`

**Comment by cmd-ntrf** [comment]
Thanks for the thorough bug report.

We currently have a pull-request pending on the exact same problem, I just have found time to merge it yet. See: https://github.com/DEAP/deap/pull/179

**Comment by weixuanfu2016** [comment]
The PR #179 try to fix the issue by sorting `common_types`, but some built-in or customized object types may be unorderable types.

```
# try customized class
class bar:
    pass

class foo:
    pass

a = [bar, foo]

sorted(a) # fail here

a = [float, int] # try built-in type

sorted(a) # fail too

```

**Comment by rhiever** [comment]
@cmd-ntrf, if you agree that the proposed change that @weixuanfu2016 posted above is the better solution (as `sorted` will not work for un-orderable types), we can submit a PR with the code change to both of the `cxOnePoint` functions.

**Comment by cmd-ntrf** [comment]
I agree there is a problem with sorting classes. However, I think there is still a potential issue with iterating on the keys. 

From Python module of the week on OrderedDict: 
> A regular dict does not track the insertion order, and iterating over it produces the values in order based on how the keys are stored in the hash table, which is in turn influenced by a random value to reduce collisions.
https://pymotw.com/3/collections/ordereddict.html

When computing common_types  this way:
```common_types = [x for x in types1 if x in types2]```
the order is still arbitrary based on Python's documentation, and could potentially change between runs with the same random seed. The keys order depends on wether hash randomization is set or not, and this is unfortunately not related to `random.seed` but to the environment variable `PYTHONHASHSEED`.
https://docs.python.org/3.3/using/cmdline.html#envvar-PYTHONHASHSEED

In most cases, I agree and observe, the keys order will be the same, but we cannot guarantee it based solely on `random.seed`, so if we want to solve this problem for good, we will need to order the keys somehow.

**Comment by rhiever** [comment]
We could start them by the `__name__` property? e.g.,

```Python
class foo:
    pass

class bar:
    pass

a = [foo, bar]

sorted(a, key=lambda x: x.__name__) # there's probably a better way to access __name__
>>>[__main__.bar, __main__.foo]
```

**Comment by cmd-ntrf** [comment]
I think we can solve this by combining the making of `types2` and `common_types`. I have created a pull-request to better explain what I mean.
https://github.com/DEAP/deap/pull/191

The `common_types` order would then depend on the second tree order, which is deterministic for a specific seed. What do you think?

**Comment by weixuanfu2016** [comment]
Looks good

**Comment by rhiever** [comment]
The logic is not straightforward, but it works. Perhaps a comment explaining what's going on in the new code would help?

**Comment by cmd-ntrf** [comment]
I agree. I am not totally done with this issue. I did some more experiments this weekend because I was wondering why the issue did not surfaced earlier with the spambase example as provided with DEAP.

In the original spambase example, we only use base type provided by Python Standard Library, bool and int. The order of the keys in dictionary is related to the hash. I determined experimentally that the hash of `bool` appears to be always smaller than the hash of `int`. Therefore the order in the set did not variate.

The order of the hash is however only true for base types, or at least bool and int. For user's class, the hash can variate, even if the order of declaration of the class and their content are kept identical. This is why @weixuanfu2016 had to reimplement the whole spambase example with `Foo` and `Bar` classes.

I have not digged into Python's internals yet to confirm all of my experimental observations. I will do it before the end of the week and compose a comment explaining why we need to get into so much trouble to find types common between the two trees.

Thanks again both of you for your help on this issue, it is very useful.

**Comment by rhiever** [comment]
Glad that we can be of assistance. We've already [implemented](https://github.com/rhiever/tpot/pull/412/files) your proposed change in TPOT and will keep an eye on this issue (and the PR) to see how it is finally implemented.
